components/pattern/src/common.rs

```diff
@@ -110,56 +110,137 @@ pub trait PatternBackend: crate::private::Sealed + 'static + core::fmt::Debug {
     fn empty() -> &'static Self::Store;
 }
 
-/// Default annotation for the literal portion of a pattern.
+/// Trait implemented on collections that can produce [`TryWriteable`]s for interpolation.
 ///
-/// For more information, see [`PlaceholderValueProvider`]. For an example, see [`Pattern`].
+/// This trait can add [`Part`]s for individual literals or placeholders. The implementations
+/// of this trait on standard types do not add any [`Part`]s.
 ///
-/// [`Pattern`]: crate::Pattern
-pub const PATTERN_LITERAL_PART: Part = Part {
-    category: "pattern",
-    value: "literal",
-};
-
-/// Default annotation for the placeholder portion of a pattern.
+/// # Examples
 ///
-/// For more information, see [`PlaceholderValueProvider`]. For an example, see [`Pattern`].
+/// A custom implementation that adds parts:
 ///
-/// [`Pattern`]: crate::Pattern
-pub const PATTERN_PLACEHOLDER_PART: Part = Part {
-    category: "pattern",
-    value: "placeholder",
-};
-
-/// Trait implemented on collections that can produce [`TryWriteable`]s for interpolation.
+/// ```
+/// use core::str::FromStr;
+/// use icu_pattern::Pattern;
+/// use icu_pattern::DoublePlaceholder;
+/// use icu_pattern::DoublePlaceholderKey;
+/// use icu_pattern::PlaceholderValueProvider;
+/// use writeable::adapters::WithPart;
+/// use writeable::adapters::WriteableAsTryWriteableInfallible;
+/// use writeable::assert_writeable_parts_eq;
+/// use writeable::Part;
+/// use writeable::Writeable;
+///
+/// let pattern = Pattern::<DoublePlaceholder>::try_from_str(
+///     "Hello, {0} and {1}!",
+///     Default::default(),
+/// )
+/// .unwrap();
+///
+/// struct ValuesWithParts<'a>(&'a str, &'a str);
+///
+/// const PART_PLACEHOLDER_0: Part = Part {
+///     category: "custom",
+///     value: "placeholder0",
+/// };
+/// const PART_PLACEHOLDER_1: Part = Part {
+///     category: "custom",
+///     value: "placeholder1",
+/// };
+/// const PART_LITERAL: Part = Part {
+///     category: "custom",
+///     value: "literal",
+/// };
+///
+/// impl PlaceholderValueProvider<DoublePlaceholderKey> for ValuesWithParts<'_> {
+///     type Error = core::convert::Infallible;
+///
+///     type W<'a> = WriteableAsTryWriteableInfallible<WithPart<&'a str>>
+///     where
+///         Self: 'a;
+///
+///     type L<'a, 'l> = WithPart<&'l str>
```

Comment:
Also, I really like how this solution composes cleanly with the Writeable adapters I added a few months ago like `WithPart`.